### PR TITLE
bump-formula-pr: audit formula before opening PR

### PR DESCRIPTION
**Pull request by ilovezfs**

### CODE REVIEW

**Comment by ilovezfs** [comment]
CC @UniqMartin

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 1]
```diff
@@ -107,22 +107,47 @@ def bump_formula_pr
...
+    else
+      system HOMEBREW_BREW_FILE, "audit", "--strict", formula.path
+      if $?.success?
+        rm backup_file
+      else
+        tmp = Pathname.new(Tempfile.new("").path).basename
+        flunker = HOMEBREW_TEMP/"#{formula.name}-#{tmp}.rb"
+        mv formula.path, flunker
+        mv backup_file, formula.path
+        odie "brew audit failed! See #{flunker}"
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Maybe just a `git diff` of the modified formula file followed by restoring the backed up original? Would prevent one temporary file from lying around and possibly being forgotten.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb [hunk 2]
```diff
@@ -107,22 +107,47 @@ def bump_formula_pr
       ]
     end
 
+    unless ARGV.dry_run?
+      backup_file = "#{formula.path}.orig"
+      if File.exist? backup_file
+        odie "Backup file #{backup_file} already exists!"
+      end
+      cp formula.path, "#{formula.path}.orig"
```
Comment (UniqMartin) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: I'd prefer if the backup resided in memory (`original_code = formula.path.read`). No need to check for file name collisions and no files that need to be cleaned up if anything goes wrong. And restoring is no more difficult than moving a file.

**Comment by UniqMartin** [comment]
Unlike the code comments not directly related to this PR, but maybe something you'd want to add to your list of things to change in this command: I haven't checked, but I'm pretty sure mirrors are ignored and thus mistreated, producing wrong results for the bumped formula if `mirror` lines are present.

**Comment by ilovezfs** [comment]
Indeed mirrors are currently unsupported and require manual follow-up (meaning double CI) or not using the command.

By the way, sneak preview:
```
diff --git a/Library/Homebrew/dev-cmd/bump-formula-pr.rb b/Library/Homebrew/dev-cmd/bump-formula-pr.rb
index 10df658..3436878 100644
--- a/Library/Homebrew/dev-cmd/bump-formula-pr.rb
+++ b/Library/Homebrew/dev-cmd/bump-formula-pr.rb
@@ -79,7 +79,15 @@ module Homebrew
     elsif !hash_type
       odie "#{formula}: no tag/revision specified!"
     else
-      odie "#{formula}: no url/#{hash_type} specified!"
+      rsrc = Resource.new { @url = new_url }
+      rsrc.download_strategy = CurlDownloadStrategy
+      rsrc.owner = Resource.new(formula.name)
+      rsrc_path = rsrc.fetch
+      if Utils.popen_read("/usr/bin/tar", "-tf", rsrc_path) =~ /\/.*\./
+        new_hash = rsrc_path.sha256
+      else
+        odie "#{formula}: no url/#{hash_type} specified!"
+      end
     end
 
     if ARGV.dry_run?

```

**Comment by UniqMartin** [comment]
> By the way, sneak preview:

To be honest, I'm utterly confused by the `tar` invocation in there. Would you care to explain what it does? Wouldn't it fail for anything not natively understood by `tar`, e.g. `xz` archives?

**Comment by ilovezfs** [comment]
`tar -tf` works fine with xz if xz is installed, somewhat to my surprise too:
```
Josephs-MacBook-Pro:Homebrew joe$ brew uninstall xz
Uninstalling /usr/local/Cellar/xz/5.2.2... (91 files, 1.4M)
Josephs-MacBook-Pro:Homebrew joe$ tar -tf libpng-1.6.21.tar.xz|tail -3
tar: Error opening archive: Child process exited with status 254Child process exited with status 254
Josephs-MacBook-Pro:Homebrew joe$ brew install xz
==> Downloading https://fossies.org/linux/misc/xz-5.2.2.tar.gz
Already downloaded: /Users/joe/Library/Caches/Homebrew/xz-5.2.2.tar.gz
==> ./configure --disable-silent-rules --prefix=/usr/local/Cellar/xz/5.2.2
==> make install
🍺  /usr/local/Cellar/xz/5.2.2: 91 files, 1.4M, built in 25 seconds
Josephs-MacBook-Pro:Homebrew joe$ tar -tf libpng-1.6.21.tar.xz|tail -3
libpng-1.6.21/README
libpng-1.6.21/pngbar.jpg
libpng-1.6.21/libpng-config.in
Josephs-MacBook-Pro:Homebrew joe$ 
```

The idea is just to validate the output prints at least one path rather than throwing up:
```
Josephs-MacBook-Pro:Homebrew joe$ tar -tf libpng-1.6.21.tar.xz|tail -3
libpng-1.6.21/README
libpng-1.6.21/pngbar.jpg
libpng-1.6.21/libpng-config.in
Josephs-MacBook-Pro:Homebrew joe$ dd if=/dev/random of=garbage.tar bs=512 count=1024
1024+0 records in
1024+0 records out
524288 bytes transferred in 0.036058 secs (14540150 bytes/sec)
Josephs-MacBook-Pro:Homebrew joe$ tar -tf garbage.tar 
tar: Unrecognized archive format
tar: Error exit delayed from previous errors.

```

**Comment by ilovezfs** [comment]
By the way, I think that "trick" that `tar -tf` works when xz is installed means that we can also simplify download_strategy.rb not to special-cause the handling .tar.xz files, since `tar -xf` also works fine when `xz` is installed.

**Comment by UniqMartin** [comment]
> The idea is just to validate the output prints at least one path rather than throwing up:

To check stuff like broken redirects etc.? The check will also fail for valid non-archives, though those are relatively rare. To my surprise, `tar` even recognizes ZIP archives …

> By the way, I think that "trick" that `tar -tf` works when xz is installed means that we can also simplify download_strategy.rb not to special-case the handling of .tar.xz files, since `tar -xf` also works fine when `xz` is installed.

I wonder when this was changed. It hasn't been too long ago when Misty added some code to explicitly pass `-z` or `-j` flags to `tar` to appease legacy OS X systems. This might also be an argument to not rely on the newer behavior, as otherwise we'll end up with two code paths for the same thing.

**Comment by ilovezfs** [comment]
Correct, it's intended to err on the side of false positives.

As for when the tar thing "changed," I'm on 10.8, so it wasn't recently.

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 2 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: 👍

File: Library/Homebrew/dev-cmd/bump-formula-pr.rb (hunk 1 above)
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/bump-formula-pr.rb]: Maybe worth autorunning `brew style --fix`

**Comment by MikeMcQuaid** [comment]
Don't have strong feelings but personally not a big fan of this; doesn't feel like it _needs_ to be in this file.

**Comment by ilovezfs** [comment]
>Don't have strong feelings but personally not a big fan of this; doesn't feel like it needs to be in this file.

What are you referring to?

**Comment by MikeMcQuaid** [comment]
> What are you referring to?

Audit being done by this script.

**Comment by ilovezfs** [comment]
Ah, I see. For what it's worth, it's proven very helpful to me so far: https://github.com/Homebrew/homebrew-core/issues/2184 all of which would have been failed CI without this.

**Comment by ilovezfs** [comment]
Of course, those issues are all preexisting problems in the formulae, not regressions introduced by bump-formula-pr, so it would work to do `brew $f && brew bump-formula-pr ...`, which is slightly less convenient, but has the upside of failing a tad faster.

**Comment by MikeMcQuaid** [comment]
@ilovezfs Yeh, I guess I'm thinking we'd typically still want a bump even if the `audit` is failing, particularly from non-maintainers who may not know how to fix the `audit` failure.

**Comment by ilovezfs** [comment]
Yes, that's a good point, and would point towards maybe making it opt-in (`--audit`), not always/opt-out (`--no-audit`)?

I'd just like to avoid adding redundant logic (meaning things `audit` knows how to do already) to `bump-formula-pr` as was being proposed here: https://github.com/Homebrew/brew/pull/372#issuecomment-226794371

**Comment by MikeMcQuaid** [comment]
@ilovezfs Opt-in sounds good if you find it useful 👍

**Comment by ilovezfs** [comment]
Cool

**Comment by MikeMcQuaid** [comment]
@ilovezfs Can you make this opt-in and merge it in? Thanks!

**Comment by MikeMcQuaid** [comment]
@ilovezfs Gentle ping on this.

**Comment by ilovezfs** [comment]
@MikeMcQuaid PR refreshed.

**Comment by MikeMcQuaid** [comment]
👍 :shipit:

**Pull request merged**